### PR TITLE
Fix main progress bar counter when `val_check_interval=int` and `check_val_every_n_epoch=None`

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

1. Add docs after #11993
2. Fix the main progress bar for these cases
3. consider total batches processed instead of `_batches_that_stepped` for validation check.

point 3 was found during review in the PR, so I included that change here since it's not a big one. Let me know if it needs another PR. Will pull it out.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7 @carmocca @awaelchli @akihironitta

### CODE REVIEW

File: docs/source/common/trainer.rst [hunk 1]
```diff
@@ -1532,7 +1532,8 @@ How often within one training epoch to check the validation set.
 Can specify as float or int.
 
 - pass a ``float`` in the range [0.0, 1.0] to check after a fraction of the training epoch.
-- pass an ``int`` to check after a fixed number of training batches.
+- pass an ``int`` to check after a fixed number of training batches. An ``int`` value can only be higher than the number of training
+  batches when ``check_val_every_n_epoch=None``, which validates after every ``N`` training batches across complete training.
```
Comment (awaelchli) [inline docs/source/common/trainer.rst:1536]: I'm unsure whether the formulation "across complete training" is correct.

Comment (akihironitta) [inline docs/source/common/trainer.rst:1536]: Maybe something like this?

```suggestion
  batches when ``check_val_every_n_epoch=None``, which validates every ``N`` training batches across epochs.
```

File: docs/source/common/trainer.rst [hunk 2]
```diff
@@ -1542,10 +1543,13 @@ Can specify as float or int.
     # check validation set 4 times during a training epoch
     trainer = Trainer(val_check_interval=0.25)
 
-    # check validation set every 1000 training batches
+    # check validation set every 1000 training batches in the current epoch
+    trainer = Trainer(val_check_interval=1000)
+
+    # check validation set every 1000 training batches across complete training
```
Comment (akihironitta) [inline docs/source/common/trainer.rst:1549]:
```suggestion
    # check validation set every 1000 training batches across epochs
```

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -405,7 +405,8 @@ def __init__(
             val_check_interval: How often to check the validation set. Pass a ``float`` in the range [0.0, 1.0] to check
                 after a fraction of the training epoch. Pass an ``int`` to check after a fixed number of training
                 batches. An ``int`` value can only be higher than the number of training batches when
-                ``check_val_every_n_epoch=None``.
+                ``check_val_every_n_epoch=None``, which validates after every ``N`` training batches across complete
+                training.
```
Comment (akihironitta) [inline pytorch_lightning/trainer/trainer.py:409]:
```suggestion
                ``check_val_every_n_epoch=None``, which validates every ``N`` training batches across epochs.
```

**Review by akihironitta — APPROVED** [review]
LGTM!

File: docs/source/common/trainer.rst (hunk 1 above)
Comment (rohitgr7) [reply docs/source/common/trainer.rst:1536]: across epochs might be incorrect if we consider max_steps too. Any better recommendation?

Also PR is incomplete right now. The progress bars are broken for some edge cases. Need to update them.

Comment (carmocca) [reply docs/source/common/trainer.rst:1536]: Is this still blocked?

File: docs/source-pytorch/common/trainer.rst [hunk 4]
```diff
@@ -1479,7 +1479,8 @@ How often within one training epoch to check the validation set.
 Can specify as float or int.
 
 - pass a ``float`` in the range [0.0, 1.0] to check after a fraction of the training epoch.
-- pass an ``int`` to check after a fixed number of training batches.
+- pass an ``int`` to check after a fixed number of training batches. An ``int`` value can only be higher than the number of training
+  batches when ``check_val_every_n_epoch=None``, which validates after every ``N`` training batches across ``trainer.fit``.
```
Comment (carmocca) [inline docs/source-pytorch/common/trainer.rst:1483]: IMO this is clearer

```suggestion
  batches when ``check_val_every_n_epoch=None``, which validates after every ``N`` training batches across epochs.
```

File: docs/source-pytorch/common/trainer.rst [hunk 5]
```diff
@@ -1489,10 +1490,13 @@ Can specify as float or int.
     # check validation set 4 times during a training epoch
     trainer = Trainer(val_check_interval=0.25)
 
-    # check validation set every 1000 training batches
+    # check validation set every 1000 training batches in the current epoch
+    trainer = Trainer(val_check_interval=1000)
+
+    # check validation set every 1000 training batches across complete training
```
Comment (carmocca) [inline docs/source-pytorch/common/trainer.rst:1496]:
```suggestion
    # check validation set every 1000 training batches across epochs
```

File: src/pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -394,7 +394,8 @@ def __init__(
             val_check_interval: How often to check the validation set. Pass a ``float`` in the range [0.0, 1.0] to check
                 after a fraction of the training epoch. Pass an ``int`` to check after a fixed number of training
                 batches. An ``int`` value can only be higher than the number of training batches when
-                ``check_val_every_n_epoch=None``.
+                ``check_val_every_n_epoch=None``, which validates after every ``N`` training batches
+                across ``trainer.fit``.
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/trainer.py:398]:
```suggestion
                across epochs.
```

File: tests/tests_pytorch/callbacks/progress/test_base_progress.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    )
+    model = BoringModel()
+    trainer.progress_bar_callback.setup(trainer, model)
+    trainer.strategy.connect(model)
+    trainer._data_connector.attach_data(model)
+    trainer.reset_train_dataloader()
+    trainer.reset_val_dataloader()
+    expected = [15, 25, 25, 15]
+
+    for expected_count in expected:
```
Comment (carmocca) [inline tests/tests_pytorch/callbacks/progress/test_base_progress.py:32]: nit:
```suggestion
    for count in expected:
```

File: src/pytorch_lightning/callbacks/progress/base.py [hunk 8]
```diff
@@ -172,6 +172,27 @@ def total_val_batches(self) -> Union[int, float]:
         assert self._trainer is not None
         return sum(self.trainer.num_val_batches) if self._trainer.fit_loop.epoch_loop._should_check_val_epoch() else 0
 
+    @property
+    def total_main_progress_bar_count_current_epoch(self) -> Union[int, float]:
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/progress/base.py:176]: The fact that the main progresss bar uses it does not mean it should be named after it

```suggestion
    def total_batches_current_epoch(self) -> Union[int, float]:
```

File: src/pytorch_lightning/callbacks/progress/base.py [hunk 9]
```diff
@@ -172,6 +172,27 @@ def total_val_batches(self) -> Union[int, float]:
...
+        total_train_batches = self.total_train_batches
+        total_val_batches = self.total_val_batches
+        assert self._trainer is not None
+
+        if total_train_batches != float("inf") and total_val_batches != float("inf"):
+            # val can be checked multiple times per epoch
+            val_check_batch = self.trainer.val_check_batch
+            if self.trainer.check_val_every_n_epoch is None:
+                batches_that_stepped = self.trainer.fit_loop.epoch_loop._batches_that_stepped
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/progress/base.py:185]: Why do we use batches_that_stepped instead of the number of batches processed? Batches where we don't `optimizer.step` should also be counted?

File: docs/source-pytorch/common/trainer.rst (hunk 4 above)
Comment (rohitgr7) [reply docs/source-pytorch/common/trainer.rst:1483]: what if epochs=None and the user is performing iteration-based training? It might be confusing then IMO

File: src/pytorch_lightning/callbacks/progress/base.py (hunk 9 above)
Comment (rohitgr7) [reply src/pytorch_lightning/callbacks/progress/base.py:185]: they are... `batches_that_stepped` doesn't care about `optimizer.step`.

```py
def training_step(...):
    return None

batches_that_stepped > 0
```

Comment (carmocca) [reply src/pytorch_lightning/callbacks/progress/base.py:185]: But batches where we accumulate don't

https://github.com/Lightning-AI/lightning/blob/d78698528de9af9177dd7444677c97e1b0171515/src/pytorch_lightning/loops/epoch/training_epoch_loop.py#L247

Comment (rohitgr7) [reply src/pytorch_lightning/callbacks/progress/base.py:185]: yep!

then we need to update the logic first
https://github.com/Lightning-AI/lightning/blob/d78698528de9af9177dd7444677c97e1b0171515/src/pytorch_lightning/loops/epoch/training_epoch_loop.py#L519-L522

do you agree?

Comment (rohitgr7) [reply src/pytorch_lightning/callbacks/progress/base.py:185]: okay. yeah, it makes sense.. logic should be updated too

Comment (rohitgr7) [reply src/pytorch_lightning/callbacks/progress/base.py]: done!

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
